keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,556 @@
+- Hot reload of the configuration.
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    server: https://<path-to-webserver>
+    authentication:
+        certificateAuthority: </path/to/ca.pem> # either this or certificateAuthorityData
+        clientCertificate: </path/to/cert.pem> # either this or clientCertificateData
+        clientKey: </path/to/key.pem> # either this or clientKeyData
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: Deny
+- type: Node
+- type: RBAC
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
+        authentication:
+        certificateAuthority: </path/to/ca.pem> # either this or certificateAuthorityData
+        clientCertificate: </path/to/cert.pem> # either this or clientCertificateData
+        clientKey: </path/to/key.pem> # either this or clientKeyData
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+```
+
+### User Stories (Optional)
+
+#### Story 1: Protecting installed CRDs
+
+Having certain Custom Resource Definitions available at cluster startup has been a long requested feature.
+One of the blockers for having a controller reconciling those CRDs is to have a protection mechanism for them, which can be achieved through Authorization Webhooks.
+However since `kube-apiserver` only allows specifying a single webhook, this would result in cluster administrators being not able to specify their own.
+Support for multiple webhooks would make this possible.
+Moreover, RBAC rules can't be used here since RBAC allows to one add permissions but not deny them.
+We won't be able to add restrictions on 'non-system' users only for a set of resources using RBAC; 'non-system' users here refers to anyone who shouldn't be able to update/delete the protected set of CRDs.
+
+Relevant Discussion Thread is sig-apimachinery: https://groups.google.com/g/kubernetes-sig-api-machinery/c/MBa19WTETMQ
+
+#### Story 2: Preventing unnecessarily nested webhooks
+
+Consider a system administrator who would like to apply a set of validations to certain requests before handing it off to webhooks defined using frameworks like Open Policy Agent.
+
+They would have to nested webhooks within the one added to the auth chain to have the intended effect.
+This enhancement allows the administrator to configure this behavior via a structured API and invoke the additional webhook only when relevant.
+This also allows administrators to define `onError` behavior for separate webhooks, leading to more predictable outcomes.
+
+#### Story 3: Denying requests on certain scenarios
+
+The authorizer chain should be powerful enough to deny anyone making a request if certain conditions are satisfied.
```

Comment:
including system:masters?  That's the use-case that openshift had for scoped tokens.

Reply:
Yup.